### PR TITLE
Report lost uart data.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/resources/uart_esp32.cc [hunk 1]
```diff
@@ -1016,6 +1043,11 @@ PRIMITIVE(get_control_flags) {
   FAIL(UNIMPLEMENTED);
 }
 
+PRIMITIVE(errors) {
+  ARGS(UartResource, uart)
+  return Smi::from(uart->errors());
```
Comment (kasperl) [inline src/resources/uart_esp32.cc:1053]: How do you want this to behave if the error count is outside the smi-range?

File: src/resources/uart_esp32.cc [hunk 2]
```diff
@@ -636,8 +650,13 @@ uint32 UartResourceGroup::on_event(Resource* r, word data, uint32 state) {
       reinterpret_cast<UartResource*>(r)->clear_tx_event_in_queue();
       break;
 
+    case UART_FIFO_OVF:
+    case UART_BUFFER_FULL:
+      reinterpret_cast<UartResource*>(r)->rx_buffer()->signal_dropped_data();
+      [[fallthrough]];
     default:
       state |= kErrorState;
+      reinterpret_cast<UartResource*>(r)->increment_errors();
```
Comment (kasperl) [inline src/resources/uart_esp32.cc:659]: Is wrapping around to zero useful? Maybe we should just avoid incrementing it after a certain point?

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/compiler/propagation/type_primitive_uart.cc [hunk 3]
```diff
@@ -31,6 +31,7 @@ TYPE_PRIMITIVE_ANY(read)
 TYPE_PRIMITIVE_ANY(wait_tx)
 TYPE_PRIMITIVE_ANY(set_control_flags)
 TYPE_PRIMITIVE_ANY(get_control_flags)
+TYPE_PRIMITIVE_ANY(errors)
```
[the remark this answers is not in the document]
Comment (floitsch) [reply src/compiler/propagation/type_primitive_uart.cc]: done.

File: src/resources/uart_esp32.cc (hunk 2 above)
Comment (floitsch) [reply src/resources/uart_esp32.cc:659]: I switched to a int64.
No reason to worry about corner cases.
done.

File: src/resources/uart_esp32.cc (hunk 1 above)
Comment (floitsch) [reply src/resources/uart_esp32.cc]: Changed it to `Primitive::Integer...`
done.

File: tests/hw/esp32/uart-error-test.toit [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import .test
+import .variants
+
+// Not that RX1 goes to TX2 and TX1 goes to RX2.
+RX1 ::= Variant.CURRENT.uart-error-in2
+TX1 ::= Variant.CURRENT.uart-error-out1
+
+RX2 ::= Variant.CURRENT.uart-error-in1
+TX2 ::= Variant.CURRENT.uart-error-out2
+
```
[the remark this answers is not in the document]
Comment (floitsch) [reply tests/hw/esp32/uart-error-test.toit:24]: done.
